parser.go

```diff
@@ -0,0 +1,191 @@
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromFile(fileName string) error {
+	var iniLines []string
+	var input io.Reader
+
+	file, err := os.Open(fileName)
+	if err != nil {
+		ErrCouldNotOpen = fmt.Errorf("cannot open file: %s", fileName)
+		return ErrCouldNotOpen
+	}
+	defer file.Close()
+
+	input = file
+	scanner := bufio.NewScanner(input)
+	for scanner.Scan() {
+		iniLines = append(iniLines, scanner.Text())
+	}
+	return parsedMap.parserLogic(iniLines)
+}
+
+// LoadFromString loads ini script from a string
+// Saves all lines locally into an array of strings
+func (parsedMap *Parser) LoadFromString(str string) error {
+	iniLines := strings.Split(str, "\n")
+	return parsedMap.parserLogic(iniLines)
+}
+
+func (parsedMap *Parser) parserLogic(iniLines []string) error {
+	var section, key, value string
+
+	for _, line := range iniLines {
+		if len(line) == 0 {
+			continue
+		} else if line[0] == '[' {
+			closingParaFound := false
+			for j, ch := range line {
+				if ch == ']' {
+					section = line[1:j]
+					section = strings.Trim(section, " ")
+					closingParaFound = true
+					if parsedMap.dictionary[section] == nil {
+						parsedMap.dictionary[section] = make(map[string]string)
+					}
+				}
+			}
+			if !closingParaFound {
+				ErrSectionNameMissingClosure = fmt.Errorf("section [%s] is missing closure paranthesis ']'", line[1:])
+				return ErrSectionNameMissingClosure
+			}
+			if parsedMap.dictionary == nil {
+				parsedMap.dictionary = make(map[string]map[string]string)
+			}
+
+		} else if line[0] == ']' {
+			ErrWrongParanthesisOrder = fmt.Errorf("wrong section paranthesis order, the section provided, %s, cannot start by ]", line)
+			return ErrWrongParanthesisOrder
+		} else if line[0] == ';' || line[0] == '#' || line[0] == ' ' || line[0] == '\n' || line[0] == '\t' {
+			continue
+		} else {
+			equalFound := false
+			for j, ch := range line {
+				if ch == '=' {
+					equalFound = true
+					key = line[0:j]
+					value = line[j+1:]
+					key = strings.Trim(key, " ")
+					value = strings.Trim(value, " ")
+					parsedMap.dictionary[section][key] = value
+					break
+				} else if ch == ']' {
+					ErrInvalidSectionName = fmt.Errorf("section name provided, %s, can't start with anything other than '['", line[:j+1])
+					return ErrInvalidSectionName
+				}
+			}
+			if !equalFound {
+				ErrMissingValueAssignment = fmt.Errorf("key provided, %s , is not assigned to a value, no '=' found", line)
+				return ErrMissingValueAssignment
+			}
+		}
+	}
+	return nil
+}
+
+// GetSectionNames provides section names of parsed file/string
+func (parsedMap *Parser) GetSectionNames() []string {
+	var sectionNames []string
+	for section := range parsedMap.dictionary {
+		sectionNames = append(sectionNames, section)
+	}
+	return sectionNames
+}
+
+// GetSections provides the dictionary/map structure
+func (parsedMap *Parser) GetSections() map[string]map[string]string {
+	return parsedMap.dictionary
```

Comment:
return a copy of the parsedData, so that any changes to it won't affect the state of the parser

Reply:
> return a copy of the parsedData, so that any changes to it won't affect the state of the parser

Done